server/src/main/kotlin/io/provenance/objectstore/gateway/service/ObjectService.kt

```diff
@@ -0,0 +1,66 @@
+package io.provenance.objectstore.gateway.service
+
+import io.provenance.objectstore.gateway.configuration.ProvenanceProperties
+import io.provenance.objectstore.gateway.exception.AccessDeniedException
+import io.provenance.objectstore.gateway.repository.ObjectPermissionsRepository
+import io.provenance.scope.encryption.model.KeyRef
+import io.provenance.scope.encryption.util.getAddress
+import io.provenance.scope.objectstore.client.CachedOsClient
+import io.provenance.scope.objectstore.util.base64Decode
+import io.provenance.scope.objectstore.util.toHex
+import io.provenance.scope.util.NotFoundException
+import io.provenance.scope.util.base64String
+import org.springframework.stereotype.Component
+import java.io.ByteArrayInputStream
+import java.security.PublicKey
+
+@Component
+class ObjectService(
+    private val objectStoreClient: CachedOsClient,
+    private val masterKey: KeyRef,
+    private val objectPermissionsRepository: ObjectPermissionsRepository,
+    private val provenanceProperties: ProvenanceProperties,
+) {
+    companion object {
+        const val OBJECT_META_TYPE_KEY = "object_type"
+    }
+
+    fun putObject(objectBytes: ByteArray, type: String?, requesterPublicKey: PublicKey): String {
+        val metadata = if (type != null) {
+            mapOf(OBJECT_META_TYPE_KEY to type)
+        } else mapOf()
+
+        return objectStoreClient.osClient.put(
```

Comment:
This might be another good reason to (someday) add a "has object for public key" route into object store.  Pre-check

Reply:
agreed